src/ImageSharp/MetaData/Profiles/ICC/IccProfile.cs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) James Jackson-South and contributors.
+// Licensed under the Apache License, Version 2.0.
+// </copyright>
+
+namespace ImageSharp
+{
+    using System;
+    using System.Collections.Generic;
+#if !NETSTANDARD1_1
+    using System.Security.Cryptography;
+#endif
+
+    /// <summary>
+    /// Represents an ICC profile
+    /// </summary>
+    public sealed class IccProfile
+    {
+        /// <summary>
+        /// The byte array to read the ICC profile from
+        /// </summary>
+        private byte[] data;
+
+        /// <summary>
+        /// The backing file for the <see cref="Entries"/> property
+        /// </summary>
+        private List<IccTagDataEntry> entries;
+
+        /// <summary>
+        /// ICC profile header
+        /// </summary>
+        private IccProfileHeader header;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IccProfile"/> class.
+        /// </summary>
+        public IccProfile()
+            : this((byte[])null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IccProfile"/> class.
+        /// </summary>
+        /// <param name="data">The raw ICC profile data</param>
+        public IccProfile(byte[] data)
+        {
+            this.data = data;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IccProfile"/> class
+        /// by making a copy from another ICC profile.
+        /// </summary>
+        /// <param name="other">The other ICC profile, where the clone should be made from.</param>
+        /// <exception cref="System.ArgumentNullException"><paramref name="other"/> is null.</exception>>
+        public IccProfile(IccProfile other)
+        {
+            Guard.NotNull(other, nameof(other));
+
+            // TODO: Do we need to copy anything else?
+            this.data = other.data;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IccProfile"/> class.
+        /// </summary>
+        /// <param name="header">The profile header</param>
+        /// <param name="entries">The actual profile data</param>
+        internal IccProfile(IccProfileHeader header, IEnumerable<IccTagDataEntry> entries)
+        {
+            Guard.NotNull(header, nameof(header));
+            Guard.NotNull(entries, nameof(entries));
+
+            this.header = header;
+            this.entries = new List<IccTagDataEntry>(entries);
+        }
+
+        /// <summary>
+        /// Gets or sets the profile header
+        /// </summary>
+        public IccProfileHeader Header
+        {
+            get
+            {
+                this.InitializeHeader();
+                return this.header;
+            }
+
+            set => this.header = value;
+        }
+
+        /// <summary>
+        /// Gets the actual profile data
+        /// </summary>
+        public List<IccTagDataEntry> Entries
```

Comment:
The API is actually very similar to the EXIF profile API. If we don't lazy initialize the properties we could potentially really hurt decoding performance. It should never be slow enough to negatively affect the user once we are past that point though.

I've always assumed we want this to be writable so this looks correct to me.